File: Lab/my_soluzioni/es0/java/Estensione/produttore.java
import java.io.BufferedReader;
import java.io.PrintWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;
import java.util.StringTokenizer;

// Produttore NON e' un filtro
public class produttore{
	public static void main(String[] args) {		
		PrintWriter[] fout = new PrintWriter[args.length];
		BufferedReader in = null;
		String riga;
		StringTokenizer tokenizer;
		int i, indiceFile;
		
		if (args.length < 1){
			System.out.println("Utilizzo: produttore <inputFilename>");
			System.exit(0);
		}
		
		in = new BufferedReader(new InputStreamReader(System.in));
		
		try {
			//apro tutti i file
			for(i=0; i<args.length; i++) {
				fout[i] = new PrintWriter( new FileWriter(args[i]) );
			}
			
			System.out.println("scrivi delle righe nel seguente formato -> numFile:testo; [EOF per terminare]");
			//scrivo sul file giusto
			while ((riga = in.readLine()) != null) { 
				tokenizer = new StringTokenizer(riga, ":");
				try {
					indiceFile = Integer.parseInt(tokenizer.nextToken());
					if(indiceFile > args.length || indiceFile < 1) {
						throw new NumberFormatException();
					}
					
					riga = tokenizer.nextToken();

					fout[indiceFile-1].println(riga);
				} 
				catch(NumberFormatException ex) {
					System.out.println("Errore indice riga");
				}
			}
			
			//chiudo tutti i file
			for(i=0; i<args.length; i++) {
				fout[i].close();
			}
		} catch(IOException ex){
			System.out.println("Errore di input");
			System.exit(2);
		} 
		
	
	}
}

File: Lab/my_soluzioni/es0/c/Estensione/produttore.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#define MAX_STRING_LENGTH 256

// produttore.c NON e' un filtro
int main(int argc, char *argv[]) {
    int   fd[argc], indiceFile, written, i, carattere;
    char  riga[MAX_STRING_LENGTH], buf[MAX_STRING_LENGTH];

    // controllo numero argomenti
    if (argc < 2) {
        perror(" numero di argomenti sbagliato");
        exit(1);
    }
    
    for(i=1; i<argc; i++) {
	fd[i] = open(argv[i], O_WRONLY | O_CREAT | O_TRUNC, 00640);

	if (fd < 0) {
	    perror("P0: Impossibile creare/aprire il file");
	    exit(2);
	}
    }
	
    printf("inserisci le righe [FORMATO -> numFile:riga]!\n");	
    // sarebbe comodo usare la scanf dato che supporta l'input con formato ma dopo la riga non può avere spazi :(
    // di conseguenza uso una getchar()
	
    // IPOTESI SEMPLIFICATIVA: il produttore non produce mai più di 9 file alla volta -> l'indice del file è composto da un solo carattere;
    // altrimenti dovrei fare un parsing della riga più articolato controllando quando trovo il separatore e salvandomi
    // ogni carattere del numero del file uno alla volta
    
    int contatoreLinea = 0;
    int fineLineaTrovato = 0;
    int c=-1;
    
    while ( (carattere = getchar()) != EOF) {	
    	contatoreLinea++;
    	//controllo il carattere di indice file
    	if( contatoreLinea==1  ) { 
    	    indiceFile = carattere-'0'; //sottraendo '0' converto il carattere in un int
    	    //printf("indice file: %d\n", indiceFile);
    	    if( indiceFile>argc-1 || indiceFile<1 ){	
    	    	//errore di formato
		perror(" formato della riga sbagliato, inserisci un numero di file corretto. Riprova:");
		do {c=getchar();} while (c!= '\n');	//consumo la riga sbagliata
		contatoreLinea=0;
		continue;	//
    	    }
    	}
    	else{
    	    //controllo la riga
    	    if( contatoreLinea==2 ) {
    	    	//butto via il separatore
    	    	continue;
    	    }
	    else{
	    	//riga effettiva
	    	written = write(fd[indiceFile], &carattere, 1); // uso della primitiva
		
	        if (written < 0) {
	           perror("P0: errore nella scrittura sul file");
                   exit(4);
	        }
	    	
	    }
	    
	    //dopo fine lina resetto il contatore
	    if( carattere == '\n') {
	    	contatoreLinea=0;
	    }    
	}
    }
    
    for(i=1; i<argc; i++) {
	close(fd[i]);
    }
}
